applications/StructuralMechanicsApplication/custom_constitutive/generic_small_strain_high_cycle_fatigue_law.cpp

```diff
@@ -0,0 +1,476 @@
+        noalias(predictive_stress_vector) = prod(r_constitutive_matrix, r_strain_vector);
+
+        // Initialize Plastic Parameters
+        double uniaxial_stress;
+        TConstLawIntegratorType::YieldSurfaceType::CalculateEquivalentStress(predictive_stress_vector, r_strain_vector, uniaxial_stress, rValues);
+
+        double fatigue_reduction_factor = this->GetFatigueReductionFactor();
+
+        uniaxial_stress /= fatigue_reduction_factor;  // Fatigue contribution
+        const double F = uniaxial_stress - threshold;
+
+        if (F > tolerance) { 
+            const double characteristic_length = ConstitutiveLawUtilities<VoigtSize>::CalculateCharacteristicLength(rValues.GetElementGeometry());
+            //const double characteristic_length = 0.01;
+            // This routine updates the PredictiveStress to verify the yield surf
+            TConstLawIntegratorType::IntegrateStressVector(
+                predictive_stress_vector, 
+                uniaxial_stress, 
+                damage, 
+                threshold, 
+                rValues, 
+                characteristic_length);
+            this->SetDamage(damage);
+            this->SetThreshold(uniaxial_stress);
+        }
+    }
+    double sign_factor;
+    HighCycleFatigueLawIntegrator<6>::CalculateTensionCompressionFactor(predictive_stress_vector, sign_factor);
+    Vector previous_stresses = ZeroVector(2);
+    const Vector& r_aux_stresses = this->GetPreviousStresses();
+    previous_stresses[1] = this->GetValue(UNIAXIAL_STRESS, previous_stresses[1])*sign_factor;
+    previous_stresses[0] = r_aux_stresses[1];
+    this->SetPreviousStresses(previous_stresses);
+    this->ResetCycleCounter();
+}
+/***********************************************************************************/
+/***********************************************************************************/
+
+template <class TConstLawIntegratorType>
+void GenericSmallStrainHighCycleFatigueLaw<TConstLawIntegratorType>::FinalizeSolutionStep(
+    const Properties& rMaterialProperties,
+    const GeometryType &rElementGeometry,
+    const Vector& rShapeFunctionsValues,
+    const ProcessInfo& rCurrentProcessInfo
+    )
+{
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template <class TConstLawIntegratorType>
+bool GenericSmallStrainHighCycleFatigueLaw<TConstLawIntegratorType>::Has(const Variable<double>& rThisVariable)
+{
+    if (rThisVariable == DAMAGE || rThisVariable == UNIAXIAL_STRESS || rThisVariable == THRESHOLD) {
+        GenericSmallStrainIsotropicDamage<TConstLawIntegratorType>::Has(rThisVariable);
+    } else if (rThisVariable == FATIGUE_REDUCTION_FACTOR) {
+        return true;
+    } else if (rThisVariable == NUMBER_OF_CYCLES) {
+        return true;
+    } else {
+        return GenericSmallStrainIsotropicDamage<TConstLawIntegratorType>::Has(rThisVariable);
```

Comment:
for those cases it is common to define a typedef to the baseclass such that one
- does not have to write the full type always
- the baseclass can be changed easily